src/dataflow/src/source/mod.rs

```diff
@@ -1500,9 +1500,14 @@ fn handle_message<S: SourceReader>(
     if let Some(len) = out.len() {
         *bytes_read += len;
     }
-    let ts_cap = cap.delayed(&ts);
 
-    output.session(&ts_cap).give(Ok(SourceOutput::new(
+    // NOTE: It is **very** important that we emit with the capability that we are given. This
+    // capability is also used to emit updates about the current read offset, and we need to make
+    // sure that these updates happen at the same timestamp, such that we can have a consistent
+    // state if/when the source restarts from persisted offsets/data.
+    //
+    // TODO: This futzes with the timestamp that emitted data has, not sure that's a problem.
+    output.session(&cap).give(Ok(SourceOutput::new(
```

Comment:
I think if I understand this issue:

we're emitting bindings that look like ((source offset, assigned time), capability time, diff)

and we're minting data that looks like (data, assigned time,diff)

and as you can see the times on the 2 are different

Reply:
the change fixes this by emitting data at capability time - but we could go the other way and emit bindings at assigned time

Reply:
maybe